src/people.rs

```diff
@@ -0,0 +1,431 @@
+use crate::{context::Context, define_data_plugin};
+use std::{
+    any::{Any, TypeId},
+    cell::{RefCell, RefMut},
+    collections::HashMap,
+};
+
+// PeopleData represents each unique person in the simulation with an id ranging
+// from 0 to population - 1. Person properties are associated with a person
+// via their id.
+struct PeopleData {
+    current_population: usize,
+    properties_map: RefCell<HashMap<TypeId, Box<dyn Any>>>,
+}
+
+define_data_plugin!(
+    PeoplePlugin,
+    PeopleData,
+    PeopleData {
+        current_population: 0,
+        properties_map: RefCell::new(HashMap::new())
+    }
+);
+
+// Represents a unique person - the id refers to that person's index in the range
+// 0 to population - 1 in the PeopleData container.
+#[derive(Clone, Copy, Hash, PartialEq, Eq)]
+pub struct PersonId {
+    pub id: usize,
+}
+
+// Individual characteristics or states related to a person, such as age or
+// disease status, are represented as "person properties". These properties
+// * are represented by a struct type that implements the PersonProperty trait,
+// * specify a Value type to represent the data associated with the property,
+// * specify an initializer, which returns the initial value
+// They may be defined with the define_person_property! macro.
+pub trait PersonProperty: Copy {
+    type Value: Copy;
+    fn initialize(context: &Context, person_id: PersonId) -> Self::Value;
+}
+
+/// Defines a person property with the following parameters:
+/// * `$person_property`: A name for the identifier type of the property
+/// * `$value`: The type of the property's value
+/// * `$default_or_initializer`: (Optional) A default value
+#[macro_export]
+macro_rules! define_person_property {
+    ($person_property:ident, $value:ty,
+        fn $initialize:ident($context:ident: &Context, $person:ident: PersonId) $ty:block) => {
+        #[derive(Copy, Clone)]
+        pub struct $person_property;
+
+        impl $crate::people::PersonProperty for $person_property {
+            type Value = $value;
+            fn initialize(
+                _context: &$crate::context::Context,
```

Comment:
why do these have leading _

Reply:
They're considered unused I think because it's not known what the inner block does; I could alternatively ignore the warning